src/thor/triplegbuilder.cc

```diff
@@ -415,49 +415,39 @@ void RemovePathEdges(valhalla::Location* location, const GraphId& edge_id) {
                           [&edge_id](const valhalla::Location::PathEdge& e) {
                             return e.graph_id() == edge_id;
                           });
-  if (pos == location->path_edges().end()) {
-    location->mutable_path_edges()->Clear(); // should never happen
-  } else if (location->path_edges_size() > 1) {
+  if (pos == location->path_edges().end())
+    throw std::logic_error("Could not find matching edge candidate");
+
+  if (location->path_edges_size() > 1) {
     location->mutable_path_edges()->SwapElements(0, pos - location->path_edges().begin());
     location->mutable_path_edges()->DeleteSubrange(1, location->path_edges_size() - 1);
   }
 }
 
 /**
- *
+ * Copy the subset of options::location into the tripleg::locations and remove edge candidates
+ * that werent removed during the construction of the route
  */
 void CopyLocations(TripLeg& trip_path,
                    const valhalla::Location& origin,
-                   const std::list<valhalla::Location>& throughs,
+                   const std::vector<valhalla::Location>& intermediates,
                    const valhalla::Location& dest,
                    const std::vector<PathInfo>::const_iterator path_begin,
                    const std::vector<PathInfo>::const_iterator path_end) {
   // origin
   trip_path.add_location()->CopyFrom(origin);
-  auto pe = path_begin;
-  RemovePathEdges(trip_path.mutable_location(trip_path.location_size() - 1), pe->edgeid);
-
-  // throughs
-  for (const auto& through : throughs) {
-    // copy
-    valhalla::Location* tp_through = trip_path.add_location();
-    tp_through->CopyFrom(through);
-    // id set
-    std::unordered_set<uint64_t> ids;
-    for (const auto& e : tp_through->path_edges()) {
-      ids.insert(e.graph_id());
-    }
-    // find id
-    auto found = std::find_if(pe, path_end, [&ids](const PathInfo& pi) {
-      return ids.find(pi.edgeid) != ids.end();
-    });
-    pe = found;
-    RemovePathEdges(trip_path.mutable_location(trip_path.location_size() - 1), pe->edgeid);
+  RemovePathEdges(&*trip_path.mutable_location()->rbegin(), path_begin->edgeid);
+  // intermediates
+  for (const auto& intermediate : intermediates) {
+    valhalla::Location* tp_intermediate = trip_path.add_location();
+    tp_intermediate->CopyFrom(intermediate);
+    // we can grab the right edge index in the path because we temporarily set it for trimming
+    RemovePathEdges(&*trip_path.mutable_location()->rbegin(),
+                    (path_begin + intermediate.leg_shape_index())->edgeid);
```

Comment:
this does worry me a little bit. the comment says it all, we rely on the `leg_shape_index` here but we are only ever setting it in `route_action.cc`. currently map matching for example, isnt sending intermediate locations so this isnt a problem but anyone that calls `triplegbuilder::build` with intermediate locations but without the `leg_shape_index` set will cause a potential crash. i dont think this is happening today but we should verify that and then we should add:

```suggestion
    if (!intermediate.has_leg_shape_index()) {
      throw std::logic_error("leg_shape_index not set for intermediate location");
    }
    RemovePathEdges(&*trip_path.mutable_location()->rbegin(),
                    (path_begin + intermediate.leg_shape_index())->edgeid);
```